### PR TITLE
Deploy management packages into bootstrap and management cluster when `package-based-lcm-beta` feature-flag is enabled

### LINKED ISSUE #1664: Integrate configuration of CC package into deployment
**Describe the feature request**

Integrate user configuration into CC package deployment.

As part of installing below clusterclass packages, we will need to pass user configuration to the package installation.

- tkg-clusterclass
    - tkg-clusterclass-docker
    - tkg-clusterclass-aws
    - tkg-clusterclass-azure
    - tkg-clusterclass-vsphere

**Describe alternatives you've considered**
Install clusterclass through ytt templating and applying templates directly. We would like to move toward carvel packaging stuff and hence we would like to use packaging approach.

**Affected product area (please put an X in all that apply)**

- [ ] APIs
- [ ] Addons
- [ ] CLI
- [ ] Docs
- [ ] IAM
- [ ] Installation
- [ ] Plugin
- [ ] Security
- [ ] Test and Release
- [ ] User Experience

**Additional context**

### PR DESCRIPTION
**Pull request by anujc25**
### What this PR does / why we need it

- This change deploys management packages into bootstrap and management cluster under feature-flag.
- Also installs kapp-controller to the management and bootstrap cluster to complete the e2e workflow. Note: We are using temporary solution to deploy kapp-controller directly.

- PR deploys below packages:

  - tkg.1.6.0
    - tkg-clusterclass.1.6.0
        - tkg-clusterclass-docker.1.6.0
        - tkg-clusterclass-aws.1.6.0
        - tkg-clusterclass-azure.1.6.0
        - tkg-clusterclass-vsphere.1.6.0

### Which issue(s) this PR fixes

Fixes #1664, #1665, #1666

### Release note

```release-note
Deploy management packages into bootstrap and management cluster under feature-flag
```

### PR Checklist

- [x] Squash the commits into one or a small number of logical commits
      
- [x] Use good commit [messages](https://github.com/vmware-tanzu/tanzu-framework/blob/main/CONTRIBUTING.md)
- [x] Ensure PR contains terms all contributors can understand and links all contributors can access

### CODE REVIEW

File: pkg/v1/config/clientconfig.go [hunk 1]
```diff
@@ -46,6 +46,9 @@ const (
 	// AVI_MANAGEMENT_CLUSTER_CONTROL_PLANE_VIP_NETWORK_NAME and AVI_MANAGEMENT_CLUSTER_CONTROL_PLANE_VIP_NETWORK_CIDR
 	// when creating a cluster.
 	FeatureFlagManagementClusterNetworkSeparation = "features.management-cluster.network-separation-beta"
+	// ClusterClass feature flag determines whether to use clusterclass based management and workload
+	// cluster provisioning or legacy way of cluster provisioning
+	FeatureFlagClusterClass = "features.global.use-clusterclass-beta"
```
Comment (vijaykatam) [inline pkg/v1/config/clientconfig.go:51]: Question: Are we bundling all of unified tkg work under this feature flag?

File: pkg/v1/tkg/client/management_components.go [hunk 2]
```diff
@@ -0,0 +1,121 @@
+// Copyright 2021 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package client
+
+import (
+	"os"
+
```
Comment (vijaykatam) [inline pkg/v1/tkg/client/management_components.go:8]: organize imports into 3 groups (std, external, local)

File: pkg/v1/tkg/client/management_components.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+		return err
+	}
+
+	managementcomponentsInstallOptions := managementcomponents.ManagementComponentsInstallOptions{
+		ClusterOptions: managementcomponents.ClusterOptions{
+			Kubeconfig:  kubeconfig,
+			Kubecontext: kubecontext,
+		},
+		KappControllerOptions: managementcomponents.KappControllerOptions{
+			KappControllerConfigFile:       "https://github.com/vmware-tanzu/carvel-kapp-controller/releases/download/v0.31.0/release.yml",
```
Comment (vijaykatam) [inline pkg/v1/tkg/client/management_components.go:42]: Could we log an issue to clean this up to get the manifest from a template? kapp-controller has a bunch of configs specific to TKG.

File: pkg/v1/tkg/managementcomponents/helper.go [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+import (
+	"gopkg.in/yaml.v3"
+
+	"github.com/vmware-tanzu/tanzu-framework/pkg/v1/tkg/constants"
+	"github.com/vmware-tanzu/tanzu-framework/pkg/v1/tkg/utils"
+)
+
+// GetTKGPackageConfigValuesFileFromUserConfig returns values file from user configuration
+func GetTKGPackageConfigValuesFileFromUserConfig(userProviderConfigValues map[string]string) (string, error) {
+	tkgPackageConfig := TKGPackageConfig{
```
[the remark this answers is not in the document]
Comment (vijaykatam) [reply pkg/v1/tkg/managementcomponents/helper.go:16]: The configs will need to mapped as well to what the actual packages expect. Don't think any of the packages currently use the upper case variables names as is.

File: pkg/v1/config/clientconfig.go (hunk 1 above)
Comment (vuil) [inline pkg/v1/config/clientconfig.go:51]: Perhaps consider generalizing the feature flag to encompass the spectrum of work involving driving cluster lvm through packages (including clusterclass packages). 
Boringly how about features.global.package-based-lcm-beta

I don't feel too strongly about it. The other alternative is to stay as is, and expect other aspects within that spectrum to be gated via another ff and deal with complications of keeping their values consistent.

File: pkg/v1/tkg/client/management_components.go [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+	}
+
+	err = utils.WriteToFile(valuesFile, configBytes)
+	if err != nil {
+		return "", err
+	}
+	return valuesFile, nil
+}
+
+func (c *TkgClient) GetUserConfigVariableValueMap(path string, rw tkgconfigreaderwriter.TKGConfigReaderWriter) (map[string]string, error) {
```
Comment (vuil) [inline pkg/v1/tkg/client/management_components.go:87]: Would be good to add a function header, explaining this is a specific implementation expecting to use a flat key-value file to provide a source of keys to filter for the valid user provided values.

File: pkg/v1/tkg/managementcomponents/helper.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (anujc25) [reply pkg/v1/tkg/managementcomponents/helper.go:23]: So, as part of MVP, I am using below schema to the `tkg` composite package values-file.
```
metadata:
  infraProvider: "aws"
configvalues:
  AWS_REGION: "us-west-2"
  CLUSTER_CIDR: ""
```

Here, the `configvalues` contains all the config values provided by the user in the raw format. The way I am imaging this, is as part of child package installation as part of `tkg` package we will be passing the required data as a secret.
See this: https://github.com/vmware-tanzu/tanzu-framework/blob/main/packages/management/tkg/bundle/config/packageinstalls/clusterclass-pi.yaml
https://github.com/vmware-tanzu/tanzu-framework/blob/main/packages/management/tkg-clusterclass/bundle/config/packageinstalls/infra-clusterclass-pi.yaml#L69

Currently, We are passing entire configvalues to the `tkg-clusterclass` child package. However, if package is accepting predefined schema for the values-file, we can write a helper function as part of the parent package, which converts this flat data values to the values schema that child package understands.

File: pkg/v1/config/clientconfig.go (hunk 1 above)
Comment (anujc25) [reply pkg/v1/config/clientconfig.go]: I think it makes sense to have common flag. I will rename the ff.

File: pkg/v1/tkg/client/management_components.go (hunk 3 above)
Comment (anujc25) [reply pkg/v1/tkg/client/management_components.go:43]: Added TODO and included issue https://github.com/vmware-tanzu/tanzu-framework/issues/1672

File: pkg/v1/config/clientconfig.go (hunk 1 above)
Comment (vijaykatam) [reply pkg/v1/config/clientconfig.go]: This name looks good.

**Review by vuil — APPROVED** [review]
Looks like a good start to some of the package-based lcm related work that we can iterate on behind a feature flag. LGTM

Nit: Please update the PR title to reflect the new feature flag name.

**Pull request merged**